src/layers/core/geojson-layer/geojson-layer.js

```diff
@@ -72,27 +72,32 @@ export default class GeoJsonLayer extends Layer {
     }
   }
 
-  _onHoverSublayer(info) {
-    this.state.pickInfos.push(info);
-  }
-
-  pick(opts) {
-    super.pick(opts);
+  _onPickSublayer(mode, info) {
+    const {pickInfos, subLayerCount} = this.state;
+    pickInfos.push(info);
 
-    const info = this.state.pickInfos.find(i => i.index >= 0);
+    if (pickInfos.length === subLayerCount) {
+      // all sublayers have been accounted for
+      let pickInfo = pickInfos.find(i => i.index >= 0) || pickInfos[0];
 
-    if (opts.mode === 'hover') {
-      this.state.pickInfos = [];
-    }
+      pickInfo = Object.assign({}, pickInfo, {
+        layer: this,
+        feature: get(pickInfo, 'object.feature') || pickInfo.object
+      });
 
-    if (!info) {
-      return;
+      switch (mode) {
+      case 'click': this.props.onClick(pickInfo); break;
+      case 'hover': this.props.onHover(pickInfo); break;
+      default: throw new Error('unknown pick type');
+      }
     }
+  }
 
-    Object.assign(opts.info, info, {
-      layer: this,
-      feature: get(info, 'object.feature') || info.object
-    });
+  pick(opts) {
+    // this is called before the onHover/onClick of sublayers
+    // pickInfo is used to collect the pick results of all sublayers
+    this.state.pickInfos.length = 0;
```

Comment:
Longer term: If this is the way picking needs to be handled for composite layers, it feels like this should be supported by the base `Layer` or by a `CompositeLayer` subclass. Looks too complicated to have to implement this in each composite layer. Let's discuss in person.